designs/component-config.md

```diff
@@ -0,0 +1,312 @@
+
+
+## Motivation
+
+This change is important because: 
+- it will help make it easier for controllers to be configured by other machine processes
+- it will reduce the required flags required to start a controller
+- allow for more configuration types which flags don't natively support
+- allow using and upgrading older configurations avoiding breaking changes in flags
+
+### Links to Open Issues
+
+- [#518 Provide a ComponentConfig to tweak the Manager](https://github.com/kubernetes-sigs/controller-runtime/issues/518)
+- [#207 Reduce command line flag boilerplate](https://github.com/kubernetes-sigs/controller-runtime/issues/207)
+- [#722 Implement ComponentConfig by default & stop using (most) flags](https://github.com/kubernetes-sigs/kubebuilder/issues/722)
+
+### Goals
+
+- Provide an interface for pulling configuration data out of exposed `ComponentConfig` types (see below for implementation)
+- Provide a new `ctrl.NewFromComponentConfig()` function for initializing a manager
+- Provide an embeddable `RuntimeConfig` type for easily authoring `ComponentConfig` types
+- Provide an `DefaultControllerConfig` to make the switch easier
+
+### Non-Goals/Future Work
+
+- `kubebuilder` implementation and design in another PR
+- Changing the default `controller-runtime` implementation
+- Dynamically reloading `ComponentConfig` object
+- Providing `flags` interface and overrides
+
+## Proposal
+
+The `ctrl.Manager` _SHOULD_ support loading configurations from `ComponentConfig` like objects. 
+An interface for that object with getters for the specific configuration parameters is created to bridge existing patterns.
+
+Without breaking the current `ctrl.NewManager` which uses an exported `ctrl.Options{}` the `manager.go` can expose a new func, `NewFromComponentConfig()` this would be able to loop through the getters to hydrate an internal `ctrl.Options{}` and pass that into `New()`.
+
+```golang
+//pkg/manager/manager.go
+
+// ManagerConfiguration defines what the ComponentConfig object for ControllerRuntime needs to support
+type ManagerConfiguration interface {
+	GetSyncPeriod() *time.Duration
+
+	GetLeaderElection() bool
+	GetLeaderElectionNamespace() string
+	GetLeaderElectionID() string
+
+	GetLeaseDuration() *time.Duration
+	GetRenewDeadline() *time.Duration
+	GetRetryPeriod() *time.Duration
+
+	GetNamespace() string
+	GetMetricsBindAddress() string
+	GetHealthProbeBindAddress() string
+
+	GetReadinessEndpointName() string
+	GetLivenessEndpointName() string
+
+	GetPort() int
+	GetHost() string
+
+	GetCertDir() string
+}
+
+func NewFromComponentConfig(config *rest.Config, scheme *runtime.Scheme, filename string, managerconfig ManagerConfiguration) (Manager, error) {
```

Comment:
Yeah, to initialize the `ctrl.Manager{}` we need to have a `config` which is currently done something like `New(config *rest.Config, options Options)` or for an implemented version:

```
mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
    Scheme:                      scheme,
    MetricsBindAddress: metricsAddr,
    LeaderElection:          enableLeaderElection,
    Port:                             9443,
    SyncPeriod:                &resyncTimeout,
})
```

Since this this is just passed through into the `Options{}` we could make the user set this on the `ManagerConfiguration` type before passing into `NewFromComponentConfig` my concern is that they we're representing non-`yaml` declarable resources into the ComponentConfig type. For example it could be:

```
mgr, err := ctrl.NewManagerFromComponentConfig(configname, &defaultv1alpha1.DefaultControllerConfiguration{
  Client: ctrl.GetConfigOrDie(),
  Scheme:  scheme, 
})
```

But given it currently separates `config` in `New(config *rest.Config, options Options) (Manager, error)`  I'm favoring not packing those fields into the `ComponentConfig` type.

thoughts?

Reply:
ComponentConfig types should only contain types that can be expressed in yaml (just like the other API definitions). They need to be serializable. Once you load them, fine to transform/copy them into runtime structures with more advanced interfaces.

Reply:
Perfect that was my understanding, so I think it makes sense to pass config and scheme since those are needed to setup the manager properly as params into this func. Agree?

Reply:
You already know which types you need in your scheme, so there isn't necessarily a reason to leave room for error there. When we parse Kubelet configs we construct the scheme internally. We have helpers like  https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/apis/config/scheme/scheme.go#L31 to avoid duplicating scheme construction boilerplate.

Reply:
Does that work because we already know the `kubeletconfig` scheme where as in this it could be custom from that specific controller, maybe I'm not just clear enough on how al that is working… 

For example in a kubebuilder scaffolded project we'd have this - https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v2/main.go#L39-L44 to register the custom scheme, and they are typically passed into the Manager like so - https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v2/main.go#L57-L63 given this currently done it seems like we'll need this internally anyways for controller-runtime to work.

So while part of the reason to include in the `NewForComponentConfig()` was to pass in for decoding from a file we also need it to init the actual Manager.

Reply:
That's fair. SGTM.